src/Features/CSharp/Portable/PullMemberUp/CSharpMembersPullerService.cs

```diff
@@ -0,0 +1,86 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    [ExportLanguageService(typeof(IMembersPullerService), LanguageNames.CSharp), Shared]
+    internal class CSharpMembersPullerService : AbstractMembersPullerService<UsingDirectiveSyntax>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpMembersPullerService()
+        {
+        }
+
+        protected override SyntaxTriviaList GetFirstMemberTrivia(SyntaxNode root)
+        {
+            var members = root switch
+            {
+                CompilationUnitSyntax compilationUnit => compilationUnit.Members,
+                NamespaceDeclarationSyntax namespaceDeclaration => namespaceDeclaration.Members,
+                _ => throw ExceptionUtilities.UnexpectedValue(root)
+            };
+
+            if (members.Count == 0)
+            {
+                return SyntaxTriviaList.Empty;
+            }
+
+            return members.First().GetLeadingTrivia();
+        }
+
+        protected override SyntaxNode EnsureLeadingTriviaBeforeFirstMember(SyntaxNode root, SyntaxTriviaList trivia)
+        {
+            var members = root switch
+            {
+                CompilationUnitSyntax compilationUnit => compilationUnit.Members,
+                NamespaceDeclarationSyntax namespaceDeclaration => namespaceDeclaration.Members,
+                _ => throw ExceptionUtilities.UnexpectedValue(root)
+            };
+
+            if (members.Count == 0)
+            {
+                return root;
+            }
+
+            var firstMember = members.First();
+            return root.ReplaceNode(firstMember, firstMember.WithLeadingTrivia(trivia));
+        }
+
+        protected override async Task<IEnumerable<UsingDirectiveSyntax>> GetImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            return root.DescendantNodesAndSelf()
+                .Where((node) => node is CompilationUnitSyntax || node is NamespaceDeclarationSyntax)
+                .SelectMany(node => node switch
+                {
+                    CompilationUnitSyntax c => c.Usings,
+                    NamespaceDeclarationSyntax n => n.Usings,
+                    _ => default,
+                })
+                .Distinct()
+                .Select(import => import.NormalizeWhitespace().WithTrailingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed));
+        }
+
+        protected override bool IsValidUnnecessaryImport(UsingDirectiveSyntax import, IEnumerable<UsingDirectiveSyntax> list)
```

Comment:
So for this part,
1. It's trying to fetch the name from the 'UsingDirectiveSyntax', you could put this logic into ISyntaxFacts.
2. 'list.Any()' is a part of logic unrelated to the language, so it would be better if it is put in the base class. Also I think it will be a little bit faster if it is an ImmutableHashSet

The general rule for writing a language service is you should try to put as many as possible codes to the base class, and for the language related part, put it into the sub class. (For your example here, the language related logic is too simple, so just put it into the ISyntaxFacts is better)

This might not be a good example because this feature doesn't have VB implementation, but we should keep this practice.